torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -0,0 +1,598 @@
+                    "parent mesh but got:\n"
+                    f"DP's global mesh: {dp_global_mesh}\n"
+                    f"TP's global mesh: {tp_global_mesh}"
+                )
+            self._global_mesh = dp_global_mesh
+            if len(self._tp_spec.placements) != 1:
+                raise NotImplementedError(
+                    f"Only supports 1D TP, not ({self._tp_spec.placements})"
+                )
+            global_placements: List[Optional[Placement]] = [None, None]
+            global_dp_mesh_dim = _mesh_resources.get_parent_mesh_dim(
+                self.mesh_info.mesh
+            )
+            global_tp_mesh_dim = _mesh_resources.get_parent_mesh_dim(self._tp_spec.mesh)
+            assert global_dp_mesh_dim is not None  # mypy
+            assert global_tp_mesh_dim is not None  # mypy
+            # TODO: Hard code inter-node FSDP + intra-node TP; need to support
+            # HSDP and DDP in place of FSDP
+            global_placements[global_dp_mesh_dim] = Shard(0)
+            global_placements[global_tp_mesh_dim] = self._tp_spec.placements[0]
+            self._global_placements = cast(
+                Tuple[Placement, ...], tuple(global_placements)
+            )
+            self._global_size = self._tp_global_size
+            self._global_stride = self._tp_global_stride
+        else:
+            if _mesh_resources.get_parent_mesh(self.mesh_info.mesh) is not None:
+                raise NotImplementedError(
+                    "For 1D DP, using a parent mesh is not supported"
+                )
+            self._global_mesh = self.mesh_info.mesh
+            self._global_placements = (Shard(0),)
+            self._global_size = param.size()
+            self._global_stride = param.stride()
+
+    @torch.no_grad()
+    def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
+        if param.device != device:
+            raise AssertionError(
+                "Expects parameter to already be moved to device "
+                f"{device} but got {param.device}"
+            )
+        param_data = param._local_tensor if isinstance(param, DTensor) else param
+        self._unsharded_size = param_data.size()
+        meta_param = torch.empty(
+            self._unsharded_size,
+            dtype=param.dtype,
+            layout=param.layout,
+            device="meta",
+            requires_grad=param.requires_grad,
+            # Assume contiguous memory layout
+        )
+        shard_rank = self.mesh_info.shard_mesh_rank
+        shard_world_size = self.mesh_info.shard_mesh_size
+        padded_param = pad_tensor_on_dim0(param_data, shard_world_size)
+        padded_chunks = torch.chunk(padded_param, shard_world_size, dim=0)
+        if self.offload_to_cpu:  # ignore `device`
+            self._sharded_param_data = param_data.new_empty(
+                padded_chunks[0].size(), device="cpu"
+            ).pin_memory()
```

Comment:
For `DTensor`, do we need to support pinning memory on CPU?

(`_sharded_param_data` is the padded tensor that is the data underlying `sharded_param`, which is the `DTensor`.)